CertiKProject/slither-certik
# Merged in tag '0.9.1'

### Notes

#### Rev1
Changes have been merged from the `0.9.1` tag of Crytic's Slither repo. This fixes some type errors which originated from within Crytic's code.

**New Developer Features**

* A printer for dominator trees
* A `--checklist` command line option to format findings as markdown
* Types now have an [`is_dynamic` method ](https://github.com/crytic/slither/pull/1175/commits/9e9f0d9c26764076293cae57cc83b3adfc1ac4bb) that determines if their values occupy a dynamically determined amount of memory
* Ability to distinguish libraries from other forms of "contracts" using the `Contract.is_library` property

**New Auditor Features**

* Checksummed addresses are [no longer reported by the `too-many-digits` detector](https://github.com/crytic/slither/issues/928)

### Testing
* Clone `slither-task` using branch `merge-0.9.1`, run `make update` and `pipenv install`, delete `utilities/slither` and clone this (branch 'merge-0.9.1' of `slither-certik`) in its place 
* Copy the [Huma Finance](https://accelerator.audit.certikpowered.info/project/09d63870-2950-11ed-9e36-994a96319390) project into the `source/huma` under the root of `slither-task`.
* Run `./run-in-pipenv.sh ./source/huma` and observe that autobuild succeeds.
* Clone [paraspace-core](https://github.com/para-space/paraspace-core/tree/0924ebdf307957723a8d1195220952b890781d2a), run `./run-in-pipenv.sh ./source/paraspace-core` and observe that autobuild succeeds
* Run `./evaluate run 100` and `./evaluate run diff` and note that the output has not changed

Slither should not crash due to a type error related to `CustomErrorContract`.

### Merge Areas

To aid reviewing, here are the sections of code where merging was performed:

In [`solc_parsing/declarations/contract.py`](https://github.com/CertiKProject/slither-certik/blob/merge-0.9.1/slither/solc_parsing/declarations/contract.py#L157-L165) we merged the following changes:
* Our change was extracting a contract's `abstract` status from the json.
* Their changes were fixing the typo `self._contract.kind` with `self._contract.contract_kind` and extracting a contract's `library` status from the json.

In the [`_convert_to_description` function in output.py](https://github.com/CertiKProject/slither-certik/blob/merge-0.9.1/slither/utils/output.py#L223-L268), we use `source_mapping` instead of `source_mapping_str`, and
in the [_convert_to_markdown function in output.py](https://github.com/CertiKProject/slither-certik/blob/merge-0.9.1/slither/utils/output.py#L223-L268), we use the `to_markdown` function to convert the source mapping to markdown rather than using a `source_mapping_to_markdown` function.

In [`output.py` in the `add_contract`](https://github.com/CertiKProject/slither-certik/blob/merge-0.9.1/slither/utils/output.py#L463-L466) function we merged the following changes:
* Our change was to add a contract's `abstract` status into the `type_specific_fields` dictionary.
* Their change was to report the `source_mapping` using its `json` form rather than passing it directly.
 
### Related Issue
https://github.com/CertiKProject/slither-task/issues/171
https://github.com/CertiKProject/slither-task/issues/132
